File: packages/android/formbricksSDK/src/main/java/com/formbricks/formbrickssdk/manager/SurveyManager.kt
package com.formbricks.formbrickssdk.manager

import android.content.Context
import com.formbricks.formbrickssdk.Formbricks
import com.formbricks.formbrickssdk.api.FormbricksApi
import com.formbricks.formbrickssdk.extensions.expiresAt
import com.formbricks.formbrickssdk.extensions.guard
import com.formbricks.formbrickssdk.logger.Logger
import com.formbricks.formbrickssdk.model.environment.EnvironmentDataHolder
import com.formbricks.formbrickssdk.model.environment.Survey
import com.formbricks.formbrickssdk.model.user.Display
import com.google.gson.Gson
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.Date
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.TimeUnit

/**
 *  The SurveyManager is responsible for managing the surveys that are displayed to the user.
 *  Filtering surveys based on the user's segments, responses, and displays.
 */
object SurveyManager {
    private const val REFRESH_STATE_ON_ERROR_TIMEOUT_IN_MINUTES = 10
    private const val FORMBRICKS_PREFS = "formbricks_prefs"
    private const val PREF_FORMBRICKS_DATA_HOLDER = "formbricksDataHolder"

    internal val refreshTimer = Timer()
    internal var displayTimer = Timer()
    internal var hasApiError = false
    internal var isShowingSurvey = false
    private val prefManager by lazy { Formbricks.applicationContext.getSharedPreferences(FORMBRICKS_PREFS, Context.MODE_PRIVATE) }
    internal var filteredSurveys: MutableList<Survey> = mutableListOf()

    private var environmentDataHolderJson: String?
        get() {
            return prefManager.getString(PREF_FORMBRICKS_DATA_HOLDER, "")
        }
        set(value) {
            if (null != value) {
                prefManager.edit().putString(PREF_FORMBRICKS_DATA_HOLDER, value).apply()
            } else {
                prefManager.edit().remove(PREF_FORMBRICKS_DATA_HOLDER).apply()
            }
        }

    private var backingEnvironmentDataHolder: EnvironmentDataHolder? = null
    var environmentDataHolder: EnvironmentDataHolder?
        get() {
            if (null != backingEnvironmentDataHolder) {
                return backingEnvironmentDataHolder
            }
            synchronized(this) {
                backingEnvironmentDataHolder = environmentDataHolderJson?.let { json ->
                    try {
                        Gson().fromJson(json, EnvironmentDataHolder::class.java)
                    } catch (e: Exception) {
                        Logger.e("Unable to retrieve environment data from the local storage.")
                        null
                    }
                }
                return backingEnvironmentDataHolder
            }
        }
        set(value) {
            synchronized(this) {
                backingEnvironmentDataHolder = value
                environmentDataHolderJson = Gson().toJson(value)
            }
        }

    /**
     * Fills up the [filteredSurveys] array
     */
    fun filterSurveys() {
        val surveys = environmentDataHolder?.data?.data?.surveys.guard { return }
        val displays = UserManager.displays ?: listOf()
        val responses = UserManager.responses ?: listOf()
        val segments = UserManager.segments ?: listOf()

        filteredSurveys = filterSurveysBasedOnDisplayType(surveys, displays, responses).toMutableList()
        filteredSurveys = filterSurveysBasedOnRecontactDays(filteredSurveys, environmentDataHolder?.data?.data?.project?.recontactDays?.toInt()).toMutableList()

        if (UserManager.userId != null) {
            if (segments.isEmpty()) {
                filteredSurveys = mutableListOf()
                return
            }

            filteredSurveys = filterSurveysBasedOnSegments(filteredSurveys, segments).toMutableList()
        }
    }

    /**
     * Checks if the environment state needs to be refreshed based on its [expiresAt] property,
     * and if so, refreshes it, starts the refresh timer, and filters the surveys.
     */
    fun refreshEnvironmentIfNeeded(force: Boolean = false) {
        if (!force) {
            environmentDataHolder?.expiresAt()?.let {
                if (it.after(Date())) {
                    Logger.d("Environment state is still valid until $it")
                    filterSurveys()
                    return
                }
            }
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                environmentDataHolder = FormbricksApi.getEnvironmentState().getOrThrow()
                startRefreshTimer(environmentDataHolder?.expiresAt())
                filterSurveys()
                hasApiError = false
            } catch (e: Exception) {
                hasApiError = true
                Logger.e("Unable to refresh environment state.")
                startErrorTimer()
            }
        }
    }

    /**
     * Checks if there are any surveys to display, based in the track action, and if so, displays the first one.
     * Handles the display percentage and the delay of the survey.
     */
    fun track(action: String) {
        val actionClasses = environmentDataHolder?.data?.data?.actionClasses ?: listOf()
        val codeActionClasses = actionClasses.filter { it.type == "code" }
        val actionClass = codeActionClasses.firstOrNull { it.key == action }
        val firstSurveyWithActionClass = filteredSurveys.firstOrNull { survey ->
            val triggers = survey.triggers ?: listOf()
            triggers.firstOrNull { it.actionClass?.name.equals(actionClass?.name) } != null
        }

        val isMultiLangSurvey = (firstSurveyWithActionClass?.languages?.size ?: 0) > 1
        if(firstSurveyWithActionClass != null && isMultiLangSurvey) {
            val currentLanguage = Formbricks.language
            val languageCode = getLanguageCode(firstSurveyWithActionClass, currentLanguage)

            if (languageCode == null) {
                Logger.e(
                    "Survey “${firstSurveyWithActionClass.name}” is not available in language “$currentLanguage”. Skipping."
                )
                return
            }

            Formbricks.setLanguage(languageCode)
        }

        val shouldDisplay = shouldDisplayBasedOnPercentage(firstSurveyWithActionClass?.displayPercentage)

        if (shouldDisplay) {
            firstSurveyWithActionClass?.id?.let {
                isShowingSurvey = true
                val timeout = firstSurveyWithActionClass.delay ?: 0.0
                stopDisplayTimer()
                displayTimer.schedule(object : TimerTask() {
                    override fun run() {
                        Formbricks.showSurvey(it)
                    }

                }, Date(System.currentTimeMillis() + timeout.toLong() * 1000))
            }
        }
    }

    private fun stopDisplayTimer() {
        try {
            displayTimer.cancel()
            displayTimer = Timer()
        } catch (_: Exception) {

        }
    }

    /**
     * Posts a survey response to the Formbricks API.
     */
    fun postResponse(surveyId: String?) {
        val id = surveyId.guard {
            Logger.e("Survey id is mandatory to set.")
            return
        }

        UserManager.onResponse(id)
    }

    /**
     * Creates a new display for the survey. It is called when the survey is displayed to the user.
     */
    fun onNewDisplay(surveyId: String?) {
        val id = surveyId.guard {
            Logger.e("Survey id is mandatory to set.")
            return
        }

        UserManager.onDisplay(id)
    }

   /**
     *  Starts a timer to refresh the environment state after the given timeout [expiresAt].
     */
    private fun startRefreshTimer(expiresAt: Date?) {
        val date = expiresAt.guard { return }
        refreshTimer.schedule(object: TimerTask() {
            override fun run() {
                Logger.d("Refreshing environment state.")
                refreshEnvironmentIfNeeded()
            }

        }, date)
    }

    /**
     *  When an error occurs, it starts a timer to refresh the environment state after the given timeout.
     */
    private fun startErrorTimer() {
        val targetDate = Date(System.currentTimeMillis() + 1000 * 60 * REFRESH_STATE_ON_ERROR_TIMEOUT_IN_MINUTES)
        refreshTimer.schedule(object: TimerTask() {
            override fun run() {
                Logger.d("Refreshing environment state after an error")
                refreshEnvironmentIfNeeded()
            }

        }, targetDate)
    }

    /**
     * Filters the surveys based on the display type and limit.
     */
    private fun filterSurveysBasedOnDisplayType(surveys: List<Survey>, displays: List<Display>, responses: List<String>): List<Survey> {
        return surveys.filter { survey ->
            when (survey.displayOption) {
                "respondMultiple" -> true

                "displayOnce" -> {
                    displays.none { it.surveyId == survey.id }
                }

                "displayMultiple" -> {
                    responses.none { it == survey.id }
                }

                "displaySome" -> {
                    survey.displayLimit?.let { limit ->
                        if (responses.any { it == survey.id }) {
                            return@filter false
                        }
                        displays.count { it.surveyId == survey.id } < limit
                    } ?: true
                }

                else -> {
                    Logger.e("Invalid Display Option")
                    false
                }
            }
        }
    }

    /**
     * Filters the surveys based on the recontact days and the [UserManager.lastDisplayedAt] date.
     */
    private fun filterSurveysBasedOnRecontactDays(surveys: List<Survey>, defaultRecontactDays: Int?): List<Survey> {
        return surveys.filter { survey ->
            val lastDisplayedAt = UserManager.lastDisplayedAt.guard { return@filter true }

            val recontactDays = survey.recontactDays ?: defaultRecontactDays

            if (recontactDays != null) {
                val daysBetween = TimeUnit.MILLISECONDS.toDays(Date().time - lastDisplayedAt.time)
                return@filter daysBetween >= recontactDays.toInt()
            }

            true
        }
    }

    /**
     * Filters the surveys based on the user's segments.
     */
    private fun filterSurveysBasedOnSegments(surveys: List<Survey>, segments: List<String>): List<Survey> {
        return surveys.filter { survey ->
            val segmentId = survey.segment?.id?.guard { return@filter false }
            segments.contains(segmentId)
        }
    }

    /**
     * Decides if the survey should be displayed based on the display percentage.
     */
    private fun shouldDisplayBasedOnPercentage(displayPercentage: Double?): Boolean {
        val percentage = displayPercentage.guard { return true }
        val randomNum = (0 until 10000).random() / 100.0
        return randomNum <= percentage
    }

    private fun getLanguageCode(survey: Survey, language: String?): String? {
        // 1) Gather all valid codes
        val availableLanguageCodes = survey.languages
            ?.map { it.language.code }
            ?: emptyList()

        // 2) No input or explicit "default" → default
        val raw = language
            ?.lowercase()
            ?.takeIf { it.isNotEmpty() }
            ?: return "default"
        if (raw == "default") return "default"

        // 3) Find matching entry by code or alias
        val selected = survey.languages
            ?.firstOrNull { entry ->
                entry.language.code.lowercase() == raw ||
                        entry.language.alias?.lowercase() == raw
            }

        // 4) If that entry is marked default → default
        if (selected?.default == true) return "default"

        // 5) If missing, disabled, or not in the available list → null
        if (selected == null
            || !selected.enabled
            || !availableLanguageCodes.contains(selected.language.code)
        ) {
            return null
        }

        // 6) Otherwise return its code
        return selected.language.code
    }
}

File: packages/ios/FormbricksSDK/FormbricksSDK/Manager/SurveyManager.swift
import SwiftUI

/// The SurveyManager is responsible for managing the surveys that are displayed to the user.
/// Filtering surveys based on the user's segments, responses, and displays.
final class SurveyManager {
    private let userManager: UserManager
    private let presentSurveyManager: PresentSurveyManager

    private init(userManager: UserManager, presentSurveyManager: PresentSurveyManager) {
        self.userManager = userManager
        self.presentSurveyManager = presentSurveyManager
    }

    static func create(userManager: UserManager, presentSurveyManager: PresentSurveyManager) -> SurveyManager {
        return SurveyManager(userManager: userManager, presentSurveyManager: presentSurveyManager)
    }

    private static let environmentResponseObjectKey = "environmentResponseObjectKey"
    internal var service = FormbricksService()
    private var backingEnvironmentResponse: EnvironmentResponse?
    /// Stores the surveys that are filtered based on the defined criteria, such as recontact days, display options etc.
    internal  private(set) var filteredSurveys: [Survey] = []
    /// Stores is a survey is being shown or the show in delayed
    internal private(set) var isShowingSurvey: Bool = false
    /// Store error state
    internal private(set) var hasApiError: Bool = false

    /// Fills up the `filteredSurveys` array
    func filterSurveys() {
        guard let environment = environmentResponse else { return }
        guard let surveys = environment.data.data.surveys else { return }

        let displays = userManager.displays ?? []
        let responses = userManager.responses ?? []
        let segments = userManager.segments ?? []

        filteredSurveys = filterSurveysBasedOnDisplayType(surveys, displays: displays, responses: responses)
        filteredSurveys = filterSurveysBasedOnRecontactDays(filteredSurveys, defaultRecontactDays: environment.data.data.project.recontactDays)

        // If we have a user, we do more filtering
        if userManager.userId != nil {
            if segments.isEmpty {
                filteredSurveys = []
                return
            }

            filteredSurveys = filterSurveysBasedOnSegments(filteredSurveys, segments: segments)
        }
    }

    /// Checks if there are any surveys to display, based in the track action, and if so, displays the first one.
    /// Handles the display percentage and the delay of the survey.
    func track(_ action: String, hiddenFields: [String: Any]? = nil) {
        guard !isShowingSurvey else { return }

        let actionClasses = environmentResponse?.data.data.actionClasses ?? []
        let codeActionClasses = actionClasses.filter { $0.type == "code" }
        let actionClass = codeActionClasses.first { $0.key == action }
        let firstSurveyWithActionClass = filteredSurveys.first { survey in
            return survey.triggers?.contains(where: { $0.actionClass?.name == actionClass?.name }) ?? false
        }

        if (firstSurveyWithActionClass == nil) {
            Formbricks.delegate?.onError(FormbricksSDKError(type: .surveyNotFoundError))
        }

        // Display percentage
        let shouldDisplay = shouldDisplayBasedOnPercentage(firstSurveyWithActionClass?.displayPercentage)
        let isMultiLangSurvey = firstSurveyWithActionClass?.languages?.count ?? 0 > 1

        if isMultiLangSurvey {
            guard let survey = firstSurveyWithActionClass else {return}
            let currentLanguage = Formbricks.language
            guard let languageCode = getLanguageCode(survey: survey, language: currentLanguage) else {
                Formbricks.logger?.error("Survey \(survey.name) is not available in language “\(currentLanguage)”. Skipping.")
                return
            }

            Formbricks.language = languageCode
        }

        // Display and delay it if needed
        if let surveyId = firstSurveyWithActionClass?.id, shouldDisplay {
            isShowingSurvey = true
            let timeout = firstSurveyWithActionClass?.delay ?? 0
            DispatchQueue.global().asyncAfter(deadline: .now() + Double(timeout)) { [weak self] in
                self?.showSurvey(withId: surveyId, hiddenFields: hiddenFields)
                Formbricks.delegate?.onSuccess(.onFoundSurvey)
            }
        }
    }
}

// MARK: - API calls -
extension SurveyManager {
    /// Checks if the environment state needs to be refreshed based on its `expiresAt` property, and if so, refreshes it, starts the refresh timer, and filters the surveys.
    func refreshEnvironmentIfNeeded(force: Bool = false,
                                    isInitial: Bool = false) {
        if (!force) {
            if let environmentResponse = environmentResponse, environmentResponse.data.expiresAt.timeIntervalSinceNow > 0 {
                Formbricks.logger?.debug("Environment state is still valid until \(environmentResponse.data.expiresAt)")
                filterSurveys()
                return
            }
        }

        service.getEnvironmentState { [weak self] result in
            switch result {
            case .success(let response):
                self?.hasApiError = false
                self?.environmentResponse = response
                self?.startRefreshTimer(expiresAt: response.data.expiresAt)
                self?.filterSurveys()
                if (isInitial) {
                    Formbricks.delegate?.onSuccess(.onFinishedSetup)
                } else {
                    Formbricks.delegate?.onSuccess(.onFinishedRefreshEnvironment)
                }
            case .failure:
                self?.hasApiError = true
                let error = FormbricksSDKError(type: .unableToRefreshEnvironment)
                Formbricks.delegate?.onError(error)
                Formbricks.logger?.error(error.message)
                self?.startErrorTimer()
            }
        }
    }

    /// Posts a survey response to the Formbricks API.
    func postResponse(surveyId: String) {
        userManager.onResponse(surveyId: surveyId)
    }

    /// Creates a new display for the survey. It is called when the survey is displayed to the user.
    func onNewDisplay(surveyId: String) {
        userManager.onDisplay(surveyId: surveyId)
        Formbricks.delegate?.onSurveyDisplayed()
    }
}

// MARK: - Present and dismiss survey window -
extension SurveyManager {
    /// Dismisses the presented survey window.
    func dismissSurveyWebView() {
        isShowingSurvey = false
        presentSurveyManager.dismissView()
    }
}

private extension SurveyManager {
    /// Presents the survey window with the given id. It is called when a survey is triggered.
    /// The survey is displayed based on the `FormbricksView`.
    /// The view controller is presented over the current context.
    func showSurvey(withId id: String, hiddenFields: [String: Any]? = nil) {
        if let environmentResponse = environmentResponse {
            presentSurveyManager.present(environmentResponse: environmentResponse, id: id, hiddenFields: hiddenFields)
        }

    }

    /// Starts a timer to refresh the environment state after the given timeout (`expiresAt`).
    func startRefreshTimer(expiresAt: Date) {
        let timeout = expiresAt.timeIntervalSinceNow
        refreshEnvironmentAfter(timeout: timeout)
    }

    /// When an error occurs, it starts a timer to refresh the environment state after the given timeout.
    func startErrorTimer() {
        refreshEnvironmentAfter(timeout: Double(Config.Environment.refreshStateOnErrorTimeoutInMinutes) * 60.0)
    }

    /// Refreshes the environment state after the given timeout.
    func refreshEnvironmentAfter(timeout: Double) {
        guard timeout > 0 else {
            return
        }

        DispatchQueue.global().asyncAfter(deadline: .now() + timeout) { [weak self] in
            Formbricks.logger?.debug("Refreshing environment state.")
            self?.refreshEnvironmentIfNeeded(force: true)
        }
    }

    /// Decides if the survey should be displayed based on the display percentage.
    func shouldDisplayBasedOnPercentage(_ displayPercentage: Double?) -> Bool {
        guard let displayPercentage = displayPercentage else { return true }
        let randomNum = Double(Int.random(in: 0..<10000)) / 100.0
        return randomNum <= displayPercentage
    }
}

// MARK: - Store data in the UserDefaults -
extension SurveyManager {
    var environmentResponse: EnvironmentResponse? {
        get {
            if let environmentResponse = backingEnvironmentResponse {
                return environmentResponse
            } else {
                if let data = UserDefaults.standard.data(forKey: SurveyManager.environmentResponseObjectKey) {
                    return try? JSONDecoder().decode(EnvironmentResponse.self, from: data)
                } else {
                    let error = FormbricksSDKError(type: .unableToRetrieveEnvironment)
                    Formbricks.delegate?.onError(error)
                    Formbricks.logger?.error(error.message)
                    return nil
                }
            }
        } set {
            if let data = try? JSONEncoder().encode(newValue) {
                UserDefaults.standard.set(data, forKey: SurveyManager.environmentResponseObjectKey)
                backingEnvironmentResponse = newValue
            } else {
                let error = FormbricksSDKError(type: .unableToPersistEnvironment)
                Formbricks.delegate?.onError(error)
                Formbricks.logger?.error(error.message)
            }
        }
    }
}

// MARK: - Helper methods -
private extension SurveyManager {
    /// Filters the surveys based on the display type and limit.
    func filterSurveysBasedOnDisplayType(_ surveys: [Survey], displays: [Display], responses: [String]) -> [Survey] {
        return surveys.filter { survey in
            switch survey.displayOption {
            case .respondMultiple:
                return true

            case .displayOnce:
                return !displays.contains { $0.surveyId == survey.id }

            case .displayMultiple:
                return !responses.contains { $0 == survey.id }

            case .displaySome:
                if let limit = survey.displayLimit {
                    if responses.contains(where: { $0 == survey.id }) {
                        return false
                    }
                    return displays.filter { $0.surveyId == survey.id }.count < limit
                } else {
                    return true
                }

            default:
                let error = FormbricksSDKError(type: .invalidDisplayOption)
                Formbricks.delegate?.onError(error)
                Formbricks.logger?.error(error.message)
                return false
            }


        }
    }

    /// Filters the surveys based on the recontact days and the `lastDisplayedAt` date.
    func filterSurveysBasedOnRecontactDays(_ surveys: [Survey], defaultRecontactDays:  Int?) -> [Survey] {
        surveys.filter { survey in
            guard let lastDisplayedAt = userManager.lastDisplayedAt else { return true }
            let recontactDays = survey.recontactDays ?? defaultRecontactDays

            if let recontactDays = recontactDays {
                return Calendar.current.numberOfDaysBetween(Date(), and: lastDisplayedAt) >= recontactDays
            }

            return true
        }
    }

    func getLanguageCode(
        survey: Survey,
        language: String?
    ) -> String? {
        // 1) Collect all codes
        let availableLanguageCodes = survey.languages?
            .map { $0.language.code }

        // 2) If no language was passed or it's the explicit "default" token → default
        guard let raw = language?.lowercased(), !raw.isEmpty else {
            return "default"
        }

        if raw == "default" {
            return "default"
        }

        // 3) Find matching entry by code or alias
        let selected = survey.languages?.first { entry in
            entry.language.code.lowercased() == raw ||
            entry.language.alias?.lowercased() == raw
        }

        // 4) If that entry is marked default → default
        if selected?.isDefault == true {
            return "default"
        }

        // 5) If no entry, or not enabled, or code not in the available list → nil
        guard
            let entry = selected,
            entry.enabled,
            availableLanguageCodes?.contains(entry.language.code) == true
        else {
            return nil
        }

        // 6) Otherwise return its code
        return entry.language.code
    }
    
    /// Filters the surveys based on the user's segments.
    func filterSurveysBasedOnSegments(_ surveys: [Survey], segments: [String]) -> [Survey] {
        return surveys.filter { survey in
            guard let segmentId = survey.segment?.id else { return false }
            return segments.contains(segmentId)
        }
    }
    
}
